### PR TITLE
feat(connector): [Noon] Add Card Payments, Capture, Void and Refund

### LINKED ISSUE #1185: feat(Connector): [Noon] Add Cards payments, Refunds, Capture and Void
### Feature Description

AddSupport for payments with Non-3DS cards via Noon.
Includes: 

- [Authorize](https://docs.noonpayments.com/payment-api/reference/initiate)
- [Capture](https://docs.noonpayments.com/payment-api/reference/capture)
- [Void](https://docs.noonpayments.com/payment-api/reference/reverse)
- [Refund](https://docs.noonpayments.com/payment-api/reference/refund)

### Possible Implementation

Add all the flows in the connector files for Noon

### Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### Are you willing to submit a PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by SamraatBansal**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Added Card Payments (3DS and Non-3DS) along with all the listed flows:

- [Authorize](https://docs.noonpayments.com/payment-api/reference/initiate)
- [Capture](https://docs.noonpayments.com/payment-api/reference/capture)
- [Void](https://docs.noonpayments.com/payment-api/reference/reverse)
- [Refund](https://docs.noonpayments.com/payment-api/reference/refund)

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Closes #1185

## How did you test it?

Unit Tests are not updated as all the tests trigger 3DS

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/noon/transformers.rs [hunk 1]
```diff
@@ -2,94 +2,185 @@ use masking::Secret;
...
+
+#[derive(Debug, Serialize, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct NoonConfiguration {
+    payment_action: NoonPaymentActions,
+    return_url: Option<String>,
 }
 
-#[derive(Default, Debug, Serialize, Eq, PartialEq)]
+#[derive(Debug, Serialize, Eq, PartialEq)]
```
Comment (ArjunKarthik) [inline crates/router/src/connector/noon/transformers.rs:42]: We don't need Eq & PartialEq in any of the structs

File: crates/router/src/connector/noon/transformers.rs [hunk 2]
```diff
@@ -2,94 +2,185 @@ use masking::Secret;
...
+        }?;
+
+        let order = NoonOrder {
+            amount: conn_utils::to_currency_base_unit(item.request.amount, item.request.currency)?,
+            currency: item.request.currency,
+            channel: NoonChannels::Web,
+            category: "pay".to_string(),
+            name: item.get_description()?,
+        };
+        let payment_action = match item.request.capture_method {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/noon/transformers.rs:98]: Use is_auto_capture method in utils.

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/noon.rs [hunk 3]
```diff
@@ -215,10 +226,19 @@ impl ConnectorIntegration<api::PSync, types::PaymentsSyncData, types::PaymentsRe
...
-        _connectors: &settings::Connectors,
+        req: &types::PaymentsSyncRouterData,
+        connectors: &settings::Connectors,
     ) -> CustomResult<String, errors::ConnectorError> {
-        Err(errors::ConnectorError::NotImplemented("get_url method".to_string()).into())
+        let connector_transaction_id = req
+            .request
+            .connector_transaction_id
+            .get_connector_transaction_id()
+            .change_context(errors::ConnectorError::MissingConnectorTransactionID)?;
```
Comment (jagan-jaya) [inline crates/router/src/connector/noon.rs:236]:
```suggestion
        let connector_transaction_id = req
            .request
            .get_connector_transaction_id()?;
```

File: crates/router/src/connector/noon/transformers.rs [hunk 4]
```diff
@@ -2,94 +2,185 @@ use masking::Secret;
...
             }),
             _ => Err(errors::ConnectorError::FailedToObtainAuthType.into()),
         }
     }
 }
-// PaymentsResponse
-//TODO: Append the remaining status flags
-#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
-#[serde(rename_all = "lowercase")]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
```
Comment (jagan-jaya) [inline crates/router/src/connector/noon/transformers.rs:138]: need both serialize and deserialize ? what about PartialEq, do you need that also?

File: crates/router/src/connector/noon/transformers.rs [hunk 5]
```diff
@@ -2,94 +2,185 @@ use masking::Secret;
...
             NoonPaymentStatus::Failed => Self::Failure,
-            NoonPaymentStatus::Processing => Self::Authorizing,
         }
     }
 }
 
-//TODO: Fill the struct with respective fields
-#[derive(Default, Debug, Clone, Serialize, Deserialize, PartialEq)]
-pub struct NoonPaymentsResponse {
+#[derive(Debug, Clone, Deserialize, PartialEq)]
```
Comment (jagan-jaya) [inline crates/router/src/connector/noon/transformers.rs:161]: PartialEq required?

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/noon/transformers.rs (hunk 5 above)
Comment (jagan-jaya) [reply crates/router/src/connector/noon/transformers.rs:161]: remove PartialEq wherever not required

**Review by ArjunKarthik — DISMISSED** [review]

File: crates/router/src/connector/noon/transformers.rs [hunk 6]
```diff
@@ -2,94 +2,188 @@ use masking::Secret;
...
 }
 
 impl From<NoonPaymentStatus> for enums::AttemptStatus {
     fn from(item: NoonPaymentStatus) -> Self {
         match item {
-            NoonPaymentStatus::Succeeded => Self::Charged,
+            NoonPaymentStatus::Authorized => Self::Authorized,
+            NoonPaymentStatus::Captured => Self::Charged,
+            NoonPaymentStatus::Reversed => Self::Voided,
+            NoonPaymentStatus::ThreeDsEnrollInitiated => Self::AuthenticationPending,
```
Comment (jagan-jaya) [inline crates/router/src/connector/noon/transformers.rs:158]: there is no pending status?

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

**Review by HyperSwitchers — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**